File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        # Rahul sir
        n = len(arr)
        
        dp = [-float('inf') for _ in range(n)]
        dp[0] = arr[0]
        
        for i in range(1, n):
            maxEle = arr[i]
            for j in range(i, max(0, i - k + 1) - 1, -1):
                maxEle = max(maxEle, arr[j])
                winSize = i - j + 1
                
                if j > 0:
                    dp[i] = max(dp[i], maxEle * winSize + dp[j - 1])
                else:
                    dp[i] = max(dp[i], maxEle * winSize)
        
        return dp[n - 1]

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        int n = arr.length;
        int[] dp = new int[n];
        dp[0] = arr[0];
        for(int i = 1; i < n; ++i) {
            int maxEle = arr[i];
            for(int j = i; j >= Math.max(0, i - k + 1); --j) {
                maxEle = Math.max(maxEle, arr[j]);
                int winSize = i - j + 1;
                if(j > 0)
                    dp[i] = Math.max(dp[i], maxEle * winSize + dp[j - 1]);
                else
                    dp[i] = Math.max(dp[i], maxEle * winSize);
            }
        }
        return dp[n - 1];
    }
}